src/meltano/cli/lock.py

```diff
@@ -0,0 +1,94 @@
+"""Lock command."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import click
+import structlog
+
+from meltano.core.plugin import PluginType
+from meltano.core.plugin_lock_service import (
+    LockfileAlreadyExistsError,
+    PluginLockService,
+)
+from meltano.core.project_plugins_service import DefinitionSource, ProjectPluginsService
+from meltano.core.tracking import CliContext, CliEvent, PluginsTrackingContext, Tracker
+
+from . import cli
+from .params import pass_project
+
+if TYPE_CHECKING:
+    from meltano.core.project import Project
+
+
+__all__ = ["lock"]
+logger = structlog.get_logger(__name__)
+
+
+@cli.command(short_help="Lock plugin definitions.")
+@click.argument(
+    "plugin-type",
+    type=click.Choice(PluginType.cli_arguments()),
+    required=False,
+)
+@click.argument("plugin_name", nargs=-1, required=False)
+@click.option("--update", "-u", is_flag=True, help="Update the lock file.")
+@pass_project()
+def lock(
+    project: Project,
+    plugin_type: str | None,
+    plugin_name: tuple[str, ...],
+    update: bool,
+):
+    """Lock plugin definitions.
+
+    \b\nRead more at https://docs.meltano.com/reference/command-line-interface#lock
+    """
+    tracker = Tracker(project)
+    tracker.add_contexts(
+        CliContext.from_command_and_kwargs(
+            "lock",
+            None,
+            update=update,
+            plugin_type=plugin_type,
+            plugin_name=plugin_name,
+        )
+    )
+    tracker.track_command_event(CliEvent.started)
+
+    lock_service = PluginLockService(project)
+    plugins_service = ProjectPluginsService(project)
+
+    with plugins_service.use_preferred_source(DefinitionSource.HUB):
+        if plugin_type:
+            plugin_type = PluginType.from_cli_argument(plugin_type)
+            plugins = plugins_service.get_plugins_of_type(plugin_type)
+            if plugin_name:
+                plugins = [plugin for plugin in plugins if plugin.name in plugin_name]
+        else:
+            # Make it a list so source preference is not lazily evaluated.
+            plugins = list(plugins_service.plugins())
+
+    tracked_plugins = []
+
+    for plugin in plugins:
+        descriptor = f"{plugin.type.descriptor} {plugin.name}"
+        tracked_plugins.append((plugin, None))
```

Comment:
@pnadolny13 Right now this is addding all matching plugins to the context (so either _all_, _all of a certain type_, or a _single one_). Do you think it's preferable to have this some other way? For example, only add plugins that were actually updated.

Reply:
Related to

- #6201

Reply:
@edgarrmondragon I want to make sure I'm understanding correctly. Are the options to either:

1. send all the plugins being considered for locking i.e. if I say "lock all" I'd get every plugin even if they all already exist and no files end up changing in my project
2. or send only plugins that actually get locked i.e. if I say "lock all" I'd get ONLY plugins that ended up causing file changes in my project

I'd probably choose option 2. Option 1 is good to know but doesnt tell us much if its every plugin every time.

@tayloramurphy what do you think?

Reply:
@pnadolny13 @edgarrmondragon I'd agree that option 2 would be best. In theory we could then understand the full set of a project's plugins vs what they had to separately lock, but I'm not sure that we'd a) prioritize that sort of analysis and b) make meaningful decisions with that distinction. So option 1 I think is fine if it's too cumbersome to do option 2.

Reply:
@pnadolny13 @tayloramurphy thanks for your input guys. I pushed the change so only plugins with updated locks are added to the context.